backend/src/utils/token.ts

```diff
@@ -0,0 +1,115 @@
+import {
+	ChatCompletionMessageParam,
+	ChatCompletionMessageToolCall,
+} from 'openai/resources/chat/completions';
+import { promptTokensEstimate, stringTokens } from 'openai-chat-tokens';
+
+import { CHAT_MODELS } from '@src/models/chat';
+import { chatGptTools } from '@src/openai';
+
+// max tokens each model can use
+const chatModelMaxTokens = {
+	[CHAT_MODELS.GPT_4_TURBO]: 128000,
+	[CHAT_MODELS.GPT_4]: 8192,
+	[CHAT_MODELS.GPT_4_0613]: 8192,
+	[CHAT_MODELS.GPT_3_5_TURBO]: 4097,
+	[CHAT_MODELS.GPT_3_5_TURBO_0613]: 4097,
+	[CHAT_MODELS.GPT_3_5_TURBO_16K]: 16385,
+	[CHAT_MODELS.GPT_3_5_TURBO_16K_0613]: 16385,
+};
+
+function countSingleToolCallTokens(toolCall: ChatCompletionMessageToolCall[]) {
+	return toolCall.reduce((acc, toolCall) => {
+		const {
+			function: { name, arguments: args },
+		} = toolCall;
+		return acc + stringTokens(name) + stringTokens(args);
+	}, 0);
+}
+
+// count total tool call in chat histroy tokens - not supported by openai-chat-tokens yet. To be removed when supported
+function countToolCallTokens(chatHistory: ChatCompletionMessageParam[]) {
```

Comment:
I don't agree with the comment. We'll still need this method that adds the tool calls across the chat history! Therefore I suggest the comment be removed.

Also I'd prefer the method to be named `estimateTOtalToolCallTokens`

Reply:
when the openai-chat-tokens library supports tool calls then they will be counted inside promptTokensEstimate - so this method will need to be removed otherwise we'll be counting them twice :)

Also can I ask why you prefer estimate? I was using countFunction to distinguish between our own functions and the estimateFunctions provided by the library

Reply:
Ahhh my bad - I misunderstood on both counts. Well justified. I guess I like estimate because it makes it clear that it's not "official", it's conjecture. But I agree it's more important to differentiate it from library methods.
No change necessary then :)